### ISSUE TITLE
Linux support

### ISSUE THREAD

**Issue by ptmt** [ptmt/react-native-desktop#59]
Qt or GTK or GTK#.

**Comment by atilacamurca** [comment]
Is there any docs how to get started in creating support for new graphical interfaces such as GTK?

**Comment by ptmt** [comment]
I've actually started to work on it (QT for now). It won't be hard to migrate to GTK later if needed. As soon as I get something I'll make the pull request to discuss.

**Comment by atilacamurca** [comment]
great, I'll be waiting.

**Comment by JohnAppleSeed-** [comment]
yes please

**Comment by cesarandreu** [comment]
@ptmt Have you considered trying to integrate with node instead of JSC?

**Comment by polarathene** [comment]
QT would be great! Wouldn't that also provide Window/OSX desktop apps too?(although I guess not as "native")

**Comment by LukasBombach** [comment]
> although I guess not as "native"

That ^^

**Comment by LukasBombach** [comment]
What's up with this? https://github.com/grassator/react-qml 
Isn't that basically a react-QT-renderer?

**Comment by ptmt** [comment]
Yeah, looks nice if you don't need react native.

**Comment by LukasBombach** [comment]
Ah right, it is only rendering QML/QT, not running natively. 👍

**Comment by ptmt** [comment]
i mean it's not the same concept where you writing Javascript for do your
UI and logic.

**Comment by kamek-pf** [comment]
I was actually wondering why you didn't target something like Qt from the get go.
It would have allowed you to target all major desktops at the same time.

Any particular reason ? genuine curiosity :p

**Comment by ptmt** [comment]
Back then it really made sense to port from UIKit -> AppKit, since it allowed to re-use 90% of the code. Of course, next-gen RN ports will consist of C++ core (which RN already has) and the UI part implemented in different libraries, including Qt.

**Comment by SuhairZain** [comment]
Hi @ptmt any updates on the status? I've been waiting for this project for more than a month now. :)

**Comment by ptmt** [comment]
It requires full-time effort or at least being the main side-project, and my patreon has failed (predictable). There is also Electron, which got unbelievable adoption and suitable solution for 90% tasks. But I believe that Linux fork will happen this year, maybe even several attempts to achieve cross-platforming. (.NET, Qt, etc.). React Native is strong as never, and community is awesome. (And I'm not saying that I've stopped, just trying to earn some money to work on open source full-time.)

**Comment by SuhairZain** [comment]
Hi @ptmt . I know the struggle. :)

I've made electron apps before but this particular one requires file system access (as in full read/write & file watchers). I didn't find a way to do that in electron since it's basically a packaged HTML and JS file and it has it's limitations, which drew me to your project. Guess I'll have to wait.

I was also thinking about lending a hand to this project. Get started on something using GTK+, but I have no experience in native desktop development whatsoever, and it would be a great help if you could point me to something that could be useful.

**Comment by kamek-pf** [comment]
Electron embeds node, you can definitely access the file system with an Electron app.

**Comment by SuhairZain** [comment]
@kamek-pf I had seen that this was possible before, but couldn't find a single example for that. Now I did another search and found this one: https://github.com/hokein/electron-sample-apps/tree/master/file-explorer. More than enough to get started and hopefully can move slowly into RN once it's released. Thanks.

**Comment by Mayzie** [comment]
@ptmt Just wondering, how is this going? Thanks :-)

**Comment by ptmt** [comment]
It has been frozen since Ubuntu announced their react-native fork.

**Comment by zserge** [comment]
In Canonical repo there seems to be no activity (in the last 7 months they only updated the readme once), also they don't seem to keep their react-native up to date.

**Comment by LukasBombach** [comment]
Also, I have tried to use it. There is no documentation and it seems very
troublesome to really use it.

**Comment by ptmt** [comment]
I can understand that, it's hard to maintain without the dedicated team working on it. It's great that Microsoft have resources to maintain Windows fork, and I'm sad that Canonical hasn't been into desktop lately. I wish I could build a really solid and useful app to attract a big sponsor to whole RN for Desktop idea. Having lightweight alternatives to Electron could be helpful in some cases, especially for small developer utilities.

**Comment by ndbroadbent** [comment]
I think react-native-windows also needs a lot more resources and a dedicated team. There's a lot of bugs and missing features, and they're struggling to keep up with the last RN releases.

I have had a pretty good experience getting my app to run on Windows 10 UWP (both phone and desktop), so I don't think there is a need for QT on Windows.

I'm about to try out react-native-macos, and like you say, it shares 90% of the code with iOS, so QT is probably unnecessary on Mac.

Now there's [react native for Ubuntu](https://github.com/CanonicalLtd/react-native), which does use QT, but it doesn't look like anyone is working on that anymore. But even if it was being actively maintained, I think it's better to use this project on Mac OS. It's always better to match the look and feel of the platform by using native controls.

> I wish I could build a really solid and useful app to attract a big sponsor to whole RN for Desktop idea.

I second this. A lot of people on HN are hating on Electron because it is so bloated, and uses a large amount of memory. I believe that React Native is a far better choice for desktop apps.

FWIW I'm working on a game that will be launched on mobile and Windows desktop, and maybe Mac if I can get react-native-macos working. Hopefully more people will start building desktop apps and games with React Native, and we can build a stronger community 🤞

**Comment by jamrizzi** [comment]
Development has started on https://github.com/jamrizzi/react-gtk. Most of the development is currently focused on the gtk bindings for node. https://github.com/jamrizzi/node-gtk3.

**Comment by LukasBombach** [comment]
Awesome!

**Comment by frol** [comment]
There is a quite interesting project here https://github.com/yue/yue ("A library for creating native cross-platform GUI apps."), which seems to try (and those attempts look quite successful to me) to implement a common framework for cross-platform native GUI with bindings to JS and LUA already available.

P.S. The author of yue seems to work in Electron team, which makes me believe that something interesting is coming...

**Comment by zserge** [comment]
@frol There has been a recent discussion on HN (https://news.ycombinator.com/item?id=15218067) and it got some rather unimpressive feedback. I would be very careful about yue (possible copyright violations, restrictions for contributors etc). And their pricing model looks bizarre for a brand new library that may or may not be trustworthy.

**Comment by beatgammit** [comment]
I just found this project and I was wonderin, are you still working on the Qt port @ptmt? If so, what state is it in and do you have a clear way for someone (e.g. me) to help out? Also, what does licensing look like?

I'm interested in building cross platform applications with a common library. With the way you're building it, could I have a dynamically linked library that can be under different licensing terms than Qt?  Writing a cross platform library seems a lot easier than writing a cross platform GUI these days...

**Comment by ptmt** [comment]
I actually never worked on Qt myself, I was doing GTK one, but not for a
long, so it's in a very early state. Canonical tried to make react-native
based on Qt, and there are some other attempts to build cross-platform GUI
frameworks, some of them may or may not use React as a core paradigm, be
sure to check all links in this thread. I strongly believe that in next
year we will see something truly great.

I'm also not a great expert in licensing, unfortunately. But probably can
help with tech if you decide to work on something like that.

**Comment by jamrizzi** [comment]
@place1 and @jamrizzi are approaching this challenge by finishing the broken and outdated node-gir (GObject Introspection) bindings for nodejs. We're then using those bindings inside of react-native-gtk. We're approaching this with the JavaScript community in mind, so we're implementing full flex support. This project is cross-platform with all operating systems. If you want to join development, hit me up on gitter.

https://github.com/Place1/node-gir
https://github.com/Place1/react-native-gtk

**Comment by polarathene** [comment]
> Canonical tried to make react-native based on Qt

Their project [got forked](https://github.com/status-im/react-native-desktop), seems to be showing active development, so you might be worth checking for anyone after Qt target(seems to be linux only for now).

**Comment by beatgammit** [comment]
> so we're implementing full flex support

Flexbox? If so, that's pretty sweet.

**Comment by jamrizzi** [comment]
Yes, we're implementing flexbox using yoga. https://github.com/facebook/yoga

**Comment by LukasBombach** [comment]
This is so cool! Totally looking forward to it!
